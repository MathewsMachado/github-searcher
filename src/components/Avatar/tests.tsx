import { render, screen } from 'utils/tests';

import { Avatar } from '.';

it('should render the passed name, username, picture and about', () => {
  render(
    <Avatar
      type='user'
      name='Mathews'
      username='machado'
      about='Github Searcher creator'
    />
  );

  expect(screen.getByText(/mathews/i)).toBeInTheDocument();
  expect(screen.getByText(/machado/i)).toBeInTheDocument();
  expect(screen.getByRole('img')).toBeInTheDocument();
  expect(screen.getByText(/creator/i)).toBeInTheDocument();
});

it('should render a default about if not passed', () => {
  render(<Avatar type='user' name='Mathews' username='machado' />);

  expect(screen.getByText(/passionate developer/i)).toBeInTheDocument();

  render(<Avatar name='Jest' username='facebook' type='repo' />);

  expect(screen.getByText(/passionate tool/i)).toBeInTheDocument();
});

it('should render the avatar as a link', () => {
  render(
    <Avatar
      type='user'
      name='Mathews'
      username='machado'
      as='a'
      href='https://github.com/machado'
    />
  );
  const link = screen.getByRole('link');

  expect(link).toHaveAttribute('href', 'https://github.com/machado');
  expect(link).toHaveStyle({ cursor: 'pointer' });
});

it('should not render the avatar username if the direction is "row" and type is "user"', () => {
  render(
    <Avatar type='user' name='Mathews' username='machado' direction='row' />
  );

  expect(screen.queryByText(/mathews/i)).toBeInTheDocument();
  expect(screen.queryByText(/machado/i)).not.toBeInTheDocument();
});

it('should render the avatar in column direction by default', () => {
  render(<Avatar type='user' name='Mathews' username='machado' />, 'wrapper');

  expect(screen.getByTestId('wrapper').firstChild).toHaveStyle({
    flexDirection: 'column',
  });
});

it('should render the avatar in row direction ', () => {
  render(
    <Avatar type='user' name='Mathews' username='machado' direction='row' />,
    'wrapper'
  );

  expect(screen.getByTestId('wrapper').firstChild).toHaveStyle({
    flexDirection: 'row',
  });
});
