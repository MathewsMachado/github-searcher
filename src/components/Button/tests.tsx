import userEvent from '@testing-library/user-event';

import { theme } from 'styles/theme';
import { render, screen } from 'utils/tests';

import { Button } from '.';

describe('<Button />', () => {
  it('should render the passed children', () => {
    render(<Button onClick={() => {}}>Search user</Button>);

    expect(screen.getByRole('button')).toHaveTextContent(/search user/i);
  });

  it('should render a black large button by default', () => {
    render(<Button onClick={() => {}}>Search user</Button>);

    expect(screen.getByRole('button')).toHaveStyle({
      backgroundColor: theme.color.primary,
    });
  });

  it('should render a pink button', () => {
    render(
      <Button color='accent' onClick={() => {}}>
        Search user
      </Button>
    );

    expect(screen.getByRole('button')).toHaveStyle({
      backgroundColor: theme.color.accent,
    });
  });

  it('should render a medium button', () => {
    render(
      <Button size='medium' onClick={() => {}}>
        Search user
      </Button>
    );

    expect(screen.getByRole('button')).toHaveStyleRule(
      'padding',
      `${theme.spacing.small} 8rem`
    );
  });

  it('should render a small button', () => {
    render(
      <Button size='small' onClick={() => {}}>
        Search user
      </Button>
    );

    expect(screen.getByRole('button')).toHaveStyleRule(
      'padding',
      `2rem ${theme.spacing.large}`
    );
  });

  it('should change color on hover', () => {
    render(<Button onClick={() => {}}>Black button</Button>);
    const blackButton = screen.getByText(/black button/i);

    expect(blackButton).toHaveStyleRule(
      'background-color',
      theme.color.primary
    );
    expect(blackButton).toHaveStyleRule('background-color', 'rgba(0,0,0,0.9)', {
      modifier: ':hover',
    });

    render(
      <Button color='accent' onClick={() => {}}>
        Pink button
      </Button>
    );
    const pinkButton = screen.getByText(/pink button/i);

    expect(pinkButton).toHaveStyleRule('background-color', theme.color.accent);
    expect(pinkButton).toHaveStyleRule(
      'background-color',
      'rgba(255,0,127,0.9)',
      { modifier: ':hover' }
    );
  });

  it('should not change the color on hover when disabled', () => {
    render(
      <Button disabled onClick={() => {}}>
        Black button
      </Button>
    );
    const blackButton = screen.getByText(/black button/i);

    expect(blackButton).toHaveStyleRule(
      'background-color',
      theme.color.primary
    );
    expect(blackButton).toHaveStyleRule(
      'background-color',
      theme.color.primary,
      { modifier: ':disabled' }
    );
    expect(blackButton).toHaveStyleRule('cursor', 'not-allowed', {
      modifier: ':disabled',
    });

    render(
      <Button disabled color='accent' onClick={() => {}}>
        Pink button
      </Button>
    );
    const pinkButton = screen.getByText(/pink button/i);

    expect(pinkButton).toHaveStyleRule('background-color', theme.color.accent);
    expect(pinkButton).toHaveStyleRule('background-color', theme.color.accent, {
      modifier: ':disabled',
    });
    expect(pinkButton).toHaveStyleRule('cursor', 'not-allowed', {
      modifier: ':disabled',
    });
  });

  it('should call the function passed to onClick prop when clicked', () => {
    const handleClick = jest.fn();

    render(<Button onClick={handleClick}>Search user</Button>);
    const button = screen.getByRole('button');

    userEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);

    userEvent.click(button);
    userEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(3);
  });
});
