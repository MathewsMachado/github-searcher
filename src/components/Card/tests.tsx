import { render, s } from 'utils/tests';

import { Card } from '.';

it('should render the passed children', () => {
  render(<Card>Github Searcher</Card>);

  expect(s.getByText(/github searcher/i)).toBeInTheDocument();
});

it('should render correctly', () => {
  render(<Card>Github Searcher</Card>, 'wrapper');

  expect(s.getByTestId(/wrapper/i).firstChild).toMatchInlineSnapshot(`
    .c0 {
      min-height: 10rem;
      padding: 1.6rem;
      background-color: #000000;
      border-radius: 1.2rem;
      -webkit-transition: all 0.1s ease-in-out;
      transition: all 0.1s ease-in-out;
    }

    .c0 > * {
      word-wrap: break-word;
    }

    .c0:hover {
      box-shadow: 0rem 1rem 1rem -0.5rem rgba(255,0,127,0.3);
    }

    .c0:active {
      box-shadow: 0rem 1.2rem 1rem -0.5rem rgba(255,0,127,0.5) inset;
    }

    <div
      class="c0"
    >
      Github Searcher
    </div>
  `);
});
