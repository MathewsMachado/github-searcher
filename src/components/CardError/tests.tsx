import { theme } from 'styles/theme';
import { render, s } from 'utils/tests';

import { CardError } from '.';

it('should render all the items correctly', () => {
  render(<CardError text='Error!' />);

  expect(s.getByText(/error/i)).toBeInTheDocument();
  expect(s.getAllByTitle(/close/i)).toHaveLength(2);
});

it('should render the items spaced-between', () => {
  render(<CardError text='Error!' />, 'wrapper');

  expect(s.getByTestId(/wrapper/i).firstChild).toHaveStyle({
    justifyContent: 'space-between',
  });
});

it('should render the text centered below medium screens and hide the icons', () => {
  render(<CardError text='Error!' />, 'wrapper');

  expect(s.getByTestId(/wrapper/i).firstChild).toHaveStyleRule(
    'justify-content',
    'center',
    { media: theme.media.breakpoints('below', 'medium') }
  );
});

it('should reduce the text size below large screens', () => {
  render(<CardError text='Error!' />);

  expect(s.getByText(/error/i)).toHaveStyleRule(
    'font-size',
    theme.font.size.xlarge,
    { media: theme.media.breakpoints('below', 'large') }
  );
});
