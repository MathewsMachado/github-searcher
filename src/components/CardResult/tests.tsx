import userEvent from '@testing-library/user-event';

import { render, s } from 'utils/tests';

import { CardResult } from '.';

it('should render all the items correctly', () => {
  render(
    <CardResult
      name='Mathews'
      username='mathewsmachado'
      type='user'
      onArrowClick={() => {}}
    />
  );

  expect(s.getByRole('img')).toBeInTheDocument();
  expect(s.getByText(/Mathews/i)).toBeInTheDocument();
  expect(s.getByTitle(/right arrow/i)).toBeInTheDocument();
  expect(s.getByText(/passionate developer/i)).toBeInTheDocument();
  expect(s.queryByText(/mathewsmachado/i)).not.toBeInTheDocument();

  render(
    <CardResult
      name='jest'
      username='facebook'
      type='repo'
      onArrowClick={() => {}}
    />
  );

  expect(s.getAllByRole('img')).toHaveLength(2);
  expect(s.getByText(/jest/i)).toBeInTheDocument();
  expect(s.getByText(/facebook/i)).toBeInTheDocument();
  expect(s.getByText(/passionate tool/i)).toBeInTheDocument();
  expect(s.getByTitle(/vs/i)).toBeInTheDocument();
  expect(s.getAllByTitle(/right arrow/i)).toHaveLength(2);
});

it('should render the VsButton based on the passed type', () => {
  render(
    <CardResult
      name='Mathews'
      username='mathewsmachado'
      type='user'
      onArrowClick={() => {}}
    />
  );

  expect(s.queryByTitle(/vs/i)).not.toBeInTheDocument();

  render(
    <CardResult
      name='jest'
      username='facebook'
      type='repo'
      onArrowClick={() => {}}
    />
  );

  expect(s.queryByTitle(/vs/i)).toBeInTheDocument();
});

it('should call the passed callback correctly', () => {
  const cb = jest.fn();

  render(
    <CardResult
      name='Mathews'
      username='mathewsmachado'
      type='user'
      onArrowClick={cb}
    />
  );
  const arrowButton = s.getByTitle(/right arrow/i);

  expect(cb).not.toBeCalled();

  userEvent.click(arrowButton);

  expect(cb).toBeCalledTimes(1);

  userEvent.click(arrowButton);
  userEvent.click(arrowButton);

  expect(cb).toBeCalledTimes(3);
});
