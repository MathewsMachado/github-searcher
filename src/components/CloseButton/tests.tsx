import { theme } from 'styles/theme';
import { render, s } from 'utils/tests';

import { CloseButton } from '.';

it('should render correctly the button children', () => {
  render(<CloseButton />);

  expect(s.getByTitle(/close/i)).toBeInTheDocument();
});

it("should have the children with the same color if it's minimal and white otherwise", () => {
  render(<CloseButton color='accent' />, 'pink');
  render(<CloseButton color='accent' minimal />, 'pinkMinimal');
  render(<CloseButton color='primary' />, 'black');
  render(<CloseButton color='primary' minimal />, 'blackMinimal');
  const [pinkButton, pinkMinimalButton, blackButton, blackMinimalButton] =
    s.getAllByRole('button');

  expect(pinkButton.firstChild).toHaveAttribute('color', theme.color.secondary);
  expect(pinkMinimalButton.firstChild).toHaveAttribute(
    'color',
    theme.color.accent
  );
  expect(blackButton.firstChild).toHaveAttribute(
    'color',
    theme.color.secondary
  );
  expect(blackMinimalButton.firstChild).toHaveAttribute(
    'color',
    theme.color.primary
  );
});
