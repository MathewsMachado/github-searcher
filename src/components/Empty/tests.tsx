import { render, s } from 'utils/tests';

import { Empty } from '.';

it('should render all the items correctly', () => {
  render(<Empty text='hello, world' />, 'wrapper');

  expect(s.getByTestId(/wrapper/i).firstChild).toHaveStyle({
    justifyContent: 'space-between',
  });
  expect(s.getByTitle(/github searcher/i)).toBeInTheDocument();
  expect(s.getByText(/hello, world/i)).toBeInTheDocument();
});
