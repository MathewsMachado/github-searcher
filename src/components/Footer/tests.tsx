import { theme } from 'app/styles';
import { render, s } from 'utils/tests';

import { Footer } from '.';

it('should render correctly all the items', () => {
  render(<Footer />);

  const text = s.getByText(/Mathews Machado, 20/i);
  const links = s.getAllByRole('link');

  expect(text).toBeInTheDocument();
  expect(links[0]).toHaveAttribute('href', expect.stringContaining('github'));
  expect(links[1]).toHaveAttribute('href', expect.stringContaining('linkedin'));
});

it('should render a black spaced-between footer', () => {
  render(<Footer />);
  const footer = s.getByRole('contentinfo');

  expect(footer).toHaveStyle({ backgroundColor: theme.color.primary });
  expect(footer).toHaveStyle({ justifyContent: 'space-between' });
});

it('should decrease the padding on small screens', () => {
  render(<Footer />);
  const footer = s.getByRole('contentinfo');

  expect(footer).toHaveStyleRule('padding', `0 ${theme.spacing.xxlarge}`);
  expect(footer).toHaveStyleRule('padding', '0 1rem', {
    media: theme.media.breakpoints('below', 'small'),
  });
});
