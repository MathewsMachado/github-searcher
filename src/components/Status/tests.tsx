import { render, screen } from 'utils/tests';

import { Status } from '.';

describe('<Status />', () => {
  it('should render a pink counter and a white label, both centralized by default', () => {
    render(<Status count={7} label='followers' />, 'wrapper');

    expect(screen.getByText(/7/i)).toBeInTheDocument();
    expect(screen.getByText(/followers/i)).toBeInTheDocument();
    expect(screen.getByTestId('wrapper').firstChild).toMatchSnapshot();
  });
});
