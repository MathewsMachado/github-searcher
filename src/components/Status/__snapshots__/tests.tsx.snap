// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Status /> should render a pink counter and a white label, both centralized by default 1`] = `
.c0 {
  text-align: center;
}

.c1 {
  display: block;
  padding-bottom: 0.8rem;
  font-size: 2.8rem;
  font-weight: 800;
  color: #FF007F;
}

.c2 {
  font-size: 1.6rem;
  font-weight: 400;
  color: #FFFFFF;
}

<div
  class="c0"
>
  <span
    class="c1"
  >
    7
  </span>
  <span
    class="c2"
  >
    followers
  </span>
</div>
`;
